Dans un nouveau notebook, exécute le code ci-dessous :

In [228]:
import json
import requests

link = "https://raw.githubusercontent.com/WildCodeSchool/java-jsonparse/master/weather.json"

# This function gets the file content from the URL
r = requests.get(link)

# You can access to the text content with this function
r.text

data = json.loads( r.text)

Affiche la valeur de l'attribut name


In [4]:
data

{'coord': {'lon': -0.13, 'lat': 51.51},
 'weather': [{'id': 300,
   'main': 'Drizzle',
   'description': 'light intensity drizzle',
   'icon': '09d'},
  {'id': 800, 'main': 'Clear', 'description': 'clear sky', 'icon': '01n'}],
 'base': 'stations',
 'main': {'temp': 280.32,
  'pressure': 1012,
  'humidity': 81,
  'temp_min': 279.15,
  'temp_max': 281.15},
 'visibility': 10000,
 'wind': {'speed': 4.1, 'deg': 80},
 'clouds': {'all': 90},
 'dt': 1485789600,
 'sys': {'type': 1,
  'id': 5091,
  'message': 0.0103,
  'country': 'GB',
  'sunrise': 1485762037,
  'sunset': 1485794875},
 'id': 2643743,
 'name': 'London',
 'cod': 200}

In [ ]:
#on print la variable unique name avec un formatage du texte pour coller à la solution
print('City name :',str(data['name']))

City name : London


Affiche les valeurs des attributs lat et lon de l'élément coord contenu dans la racine



In [ ]:
#on va parcourir le dictionnaire coord et afficher lat et lon quand on les retrouve avec un print formater pour choisir le texte
for ele, val in data['coord'].items():
    if ele=='lon':
        print("City Longitude: ",val)
    if ele=='lat':
        print("City latitude: ",val)

City Longitude:  -0.13
City latitude:  51.51


Parcours tous les éléments dans weather et affiche le contenu de main (peut-être avec une boucle ?)

In [ ]:
#on parcours le dictionnaire du data
for ele, val in data.items():
    #Quand on arrive à la partie weather on passe sur une liste puisque c'est une liste que l'on va parcourir
    if ele == 'weather':
        for index in val:
            #Chaque valeur de cette liste est un dictionnaire que l'on parcours pour retrouver l'index main
            for indice, valeur in index.items():
                if indice == 'main':
                    print(ele, ": ", valeur)

weather :  Drizzle
weather :  Clear


Voici le résultat attendu de ces 3 premières questions :

* City name: London
* City latitude: 51.51
* City longitude: -0.13
* Weather: Drizzle
* Weather: Clear

Essaye maintenant d'importer ton fichier json dans pandas :

In [223]:
import pandas as pd
pd.json_normalize(r.json())

weather      base  visibility  \
0  [{'id': 300, 'main': 'Drizzle', 'description':...  stations       10000   

           dt       id    name  cod  coord.lon  coord.lat  main.temp  ...  \
0  1485789600  2643743  London  200      -0.13      51.51     280.32  ...   

   main.temp_max  wind.speed  wind.deg  clouds.all  sys.type  sys.id  \
0         281.15         4.1        80          90         1    5091   

   sys.message  sys.country  sys.sunrise  sys.sunset  
0       0.0103           GB   1485762037  1485794875  

[1 rows x 23 columns]

Tu obtiens un DataFrame avec une seule ligne ? C'est normal, notre fichier JSON contient un premier niveau "weather". Nous allons indiquer à Pandas de descendre au niveau inférieur (ici dans l'enregistrement weather). Pour cela, nous utilisons le paramètre record_path. Nous pouvons aussi utiliser le paramètre meta pour indiquer de conserver des éléments d'un niveau supérieur:

In [224]:
pd.json_normalize(r.json(),  
                  record_path= "weather",
                  meta = 'dt')

id     main              description icon          dt
0  300  Drizzle  light intensity drizzle  09d  1485789600
1  800    Clear                clear sky  01n  1485789600

Filtre maintenant ton DataFrame avec Pandas pour afficher uniquement les colonnes suivantes : main (celui qui est à l'intérieur de weather), name, and visibility.

In [ ]:
#différents tests inutiles au final

# for ele, val in data_1.items():
#     if ele == "visibility":
#         print(val)
# pd.json_normalize(r.json(),   
#                  "weather",
#                 ['name','visibility']
                
# )
# pd.json_normalize(r.json(),   
#                 record_path= ["weather"][0],
#                 meta=[['weather'],'name','visibility', ['main','temp']]
                
# )

id     main              description icon    name visibility
0  300  Drizzle  light intensity drizzle  09d  London      10000
1  800    Clear                clear sky  01n  London      10000

In [227]:
#on récupère les informations souhaitées
df_filtre = pd.json_normalize(r.json(),
                record_path=['weather'],
                meta=['name','visibility']
                )
#on filtre le dataset récupéré
df_filtre[['main','name','visibility']]

#ou

df_filtre.loc[:,['main','name','visibility']]

main    name visibility
0  Drizzle  London      10000
1    Clear  London      10000